src/platforms/posix/px4_daemon/client.cpp

```diff
@@ -0,0 +1,327 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+/**
+ * @file client.cpp
+ *
+ * @author Julian Oes <julian@oes.ch>
+ * @author Beat Küng <beat-kueng@gmx.net>
+ */
+
+#include <stdio.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/un.h>
+
+#include <string>
+
+#include <px4_log.h>
+#include "client.h"
+
+namespace px4_daemon
+{
+
+
+namespace client
+{
+static Client *_instance;
+}
+
+Client::Client() :
+	_uuid(0),
+	_client_send_pipe_fd(-1)
+{
+	client::_instance = this;
+}
+
+Client::~Client()
+{
+	client::_instance = nullptr;
+}
+
+int
+Client::generate_uuid()
+{
+	int rand_fd = open("/dev/urandom", O_RDONLY);
+
+	if (rand_fd < 0) {
+		PX4_ERR("open urandom");
+		return rand_fd;
+	}
+
+	int rand_read = read(rand_fd, &_uuid, sizeof(_uuid));
+
+	if (rand_read != sizeof(_uuid)) {
+		PX4_ERR("rand read fail");
+		return rand_read;
+	}
+
+	close(rand_fd);
+	return 0;
+}
+
+int
+Client::process_args(const int argc, const char **argv)
+{
+	// Should be called with at least one command to forward to the server.
+	if (argc < 2) {
+		PX4_ERR("Not enough arguments.");
+		return -1;
+	}
+
+
+	// Prepare return pipe first to avoid a race.
+	int ret = _prepare_recv_pipe();
+
+	if (ret != 0) {
+		PX4_ERR("Could not prepare recv pipe");
+		return -2;
+	}
+
+	ret = _send_cmds(argc, argv);
+
+	if (ret != 0) {
+		PX4_ERR("Could not send commands");
+		return -3;
+	}
+
+	return _listen();
+}
+
+
+int
+Client::_prepare_recv_pipe()
+{
+	int ret = get_client_recv_pipe_path(_uuid, _recv_pipe_path, sizeof(_recv_pipe_path));
+
+	if (ret < 0) {
+		PX4_ERR("failed to assemble path");
+		return ret;
+	}
+
+	unlink(_recv_pipe_path);
```

Comment:
I think this is a precaution in case a client with the same uuid exists/ed? I think we have to be careful here (however unlikely the case is): will this kill the other client? If so, we might be better off trying a different uuid.